### ISSUE TITLE
Override default toString() for org.cactoos.iterator.Immutable

### ISSUE THREAD

**Issue by atapin** [yegor256/cactoos#1102]
As `org.cactoos.iterator.Immutable` provides the default `toString()` implementation, we see not friendly messages like:

```
Item #2 of org.cactoos.iterator.Immutable@6d86b085 is NULL
```

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by llorllale** [comment]
@0crat in

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/cactoos/issues/1102#issuecomment-482977066))

@llorllale Job [#1102](https://github.com/yegor256/cactoos/issues/1102) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
Bug was reported, see [§29](http://www.zerocracy.com/policy.html#29): +15 point(s) just awarded to @atapin[/z](https://www.0crat.com/u/atapin)

**Comment by 0crat** [comment]
The job [#1102](https://github.com/yegor256/cactoos/issues/1102) assigned to @scristalli[/z](https://www.0crat.com/u/scristalli), here is [why](https://www.0crat.com/footprint/C63314D6Z/30cf1190-5aca-4087-a504-16aebde875c6); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be no monetary reward for this job

**Comment by scristalli** [comment]
It is not clear to me what the expected behavior should be. If we want a display of the whole collection as for `Iterable`, then a traversal (and consequently, a copy) of the `Iterator` is required.
Whether this is the case or not, it seems to me that this problem is broader than just `Immutable`, as I couldn't find any other `Iterator` implementing this.

**Comment by scristalli** [comment]
@0crat wait

**Comment by scristalli** [comment]
Just to clarify, the impediment I see is only strictly related to the expected output, which is not clear neither from the description, nor from other `Iterator` classes in the project or standard implementations as far as I know.

**Comment by scristalli** [comment]
@0crat waiting for #1128

**Comment by 0crat** [comment]
> @0crat waiting for #1128  ([here](https://github.com/yegor256/cactoos/issues/1102#issuecomment-494913074))

@scristalli The impediment for [#1102](https://github.com/yegor256/cactoos/issues/1102) was registered successfully by @scristalli[/z](https://www.0crat.com/u/scristalli)

**Comment by llorllale** [comment]
@atapin @victornoel @scristalli it just occurred to me that making `Iterator.toString()` print out all elements will consume the iterator.

While it *is* possible to cache these values during a call to `toString()`, we would have to manage state to determine if the iterator has been *actually* exhausted during a "real" iteration (ie. in calls to `next()`) as opposed to just `toString()`. It's unexpected and messy.

Consider that [guava](https://google.github.io/guava/releases/15.0/api/docs/com/google/common/collect/Iterators.html#toString(java.util.Iterator)) suffers from the same problem. (I have not tested [Apache Commons](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/IteratorUtils.html#toString-java.util.Iterator-org.apache.commons.collections4.Transformer-java.lang.String-java.lang.String-java.lang.String-)).

**Comment by victornoel** [comment]
@llorllale the same problem exists also in jOOL for the record, but they solve it by storing the content of the stream in a buffer:
https://github.com/jOOQ/jOOL/blob/master/jOOL/src/main/java/org/jooq/lambda/SeqImpl.java#L602

**Comment by llorllale** [comment]
@victornoel yes, and take a look at their `stream()` method:

https://github.com/jOOQ/jOOL/blob/889d87c85ca57bafd4eddd78e0f7ae2804d2ee86/jOOL/src/main/java/org/jooq/lambda/SeqImpl.java#L67-L71

`SeqImpl` is a big class and that mutable `buffered` attribute is there just because they wanted a pretty `toString()`.

My other concern is how that `stream()` method behaves differently depending on whether `toString()` was called previously: 

* `SeqImpl.stream()` returns a fresh new stream if `toString()` was called previously, otherwise it returns a stream that may or may not already be consumed
* `SeqImpl.toString()` returns a string representation of the elements *if the underlying stream has not yet been consumed*

It's just a mess.

As per the [Stream Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html):

> A stream should be operated on (invoking an intermediate or terminal stream operation) only once. 

The use of "should" as opposed to "must" softens their point a little, but it should be clear that the intention is for streams to be discarded after being *consumed* - like iterators.

**Comment by victornoel** [comment]
@llorllale I think that `Iterator` implementations should implement `toString` depending on their state:
- `org.cactoos.iterator.Immutable` should delegate to its member iterator `toString` implementation
- `IteratorOf` should print out the actual elements as well as the current position since they are known
- and so on for every Iterator implementation

In some case it means that calling `toString` on Immutable won't give much information (if the wrapped iterator doesn't give some) while in some other cases, it will.

Basically, we shouldn't try to assume anything about the wrapped iterator in `Immutable`, just delegate.

**Comment by scristalli** [comment]
I agree, delegating for `Immutable` seems the right choice.
I personally dislike implementations of `toString` with side effects like the Guava one, intuitively I think that printing the state of an object should never change the state itself.
@llorllale if you agree for `Immutable` I can make a pull request and close this issue, then maybe we can move discussion for other iterators in #1128?

**Comment by llorllale** [comment]
@victornoel I wouldn't be too quick to assume that `IteratorOf` will always know the elements. A ctor could be added later on for a different source of elements (eg. IO).

**Comment by victornoel** [comment]
@llorllale that's not what I said. I just said that the encapsulated state of `org.cactoos.iterator.Immutable`, which can be any instance of `Iterator` here, knows better how to print itself, whether it is by printing its elements or just saying it comes from an IO as you illustrate.
The `IteratorOf` example was just that, an example :)

So I'm just proposing to delegate to the objects that knows better, and each implementation will do what it can.

**Comment by llorllale** [comment]
@victornoel that's not what you said. In your comment, both `Immutable` and `IteratorOf` are sibling bullet points subordinate to `Iterator`.

**Comment by victornoel** [comment]
@llorllale I've edited my comment, now is it understandable?

**Comment by llorllale** [comment]
@victornoel yes, it is understandable. And yes, I agree.

@atapin @scristalli @victornoel so, to conclude: tickets for `toString` on these iterators would have to be opened on an per-implementation basis, correct? For `Immutable`, the best we can do is delegate to the decorated iterator's `toString`. This may or may not satisfy @atapin 's implicit request in this ticket

**Comment by scristalli** [comment]
@atapin what do you think, can we close the issue? To recap, `Immutable.toString()` simply delegates to its wrapped iterator, for the reasons discussed above.

**Comment by atapin** [comment]
@scristalli @llorllale yes, let's close this issue

@0crat close this issue

**Comment by 0crat** [comment]
Are you speaking to me or about me [here](https://github.com/yegor256/cactoos/issues/1102#issuecomment-503455742); you must always start your message with my name if you want to address it to me, see [§1](http://www.zerocracy.com/policy.html#1)

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) please, pay attention to this issue

**Comment by 0crat** [comment]
@atapin[/z](https://www.0crat.com/u/atapin) this project will fix the problem faster if you donate a few dollars to it; just [click here](https://www.0crat.com/contrib/C63314D6Z) and pay via Stripe, it's very fast, convenient and appreciated; thanks a lot!

**Comment by atapin** [comment]
@0crat close the issue

**Comment by victornoel** [comment]
@atapin just close it yourself…

**Comment by 0crat** [comment]
> @0crat close the issue ([here](https://github.com/yegor256/cactoos/issues/1102#issuecomment-503455954))

@atapin Can't understand "close", try one of these:
  - `assign` Assign a performer to this issue
  - `boost` Set a boost for the job
  - `continue` Remove a job's impediment
  - `hello` Just say hello
  - `in` Register this issue in scope (WBS)
  - `out` Close the order and remove this job from scope
  - `pay` Pay a user some extra cash
  - `quality` Review a task
  - `resign` Resign from this issue
  - `status` Check the status of the job
  - `version` Print current version of the bot
  - `wait` Register an impediment for a job

**Comment by 0crat** [comment]
GitHub issue is still open, won't close

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @scristalli[/z](https://www.0crat.com/u/scristalli), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1102](https://github.com/yegor256/cactoos/issues/1102) is now out of scope

**Comment by sereshqua** [comment]
@scristalli please edit your comment that does not start with the name of the user you are referring it to: [see](https://github.com/yegor256/cactoos/issues/1102#issuecomment-496817005)

**Comment by scristalli** [comment]
@sereshqua done

**Comment by sereshqua** [comment]
@scristalli thx! it's been a lot of work in this ticket, i did not want to give just acceptable quality here :)

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by victornoel** [comment]
@scristalli 
> thx! it's been a lot of work in this ticket, i did not want to give just acceptable quality here :)

That's not how this work… the idea of QA is to reward or punish. `acceptable` is being paid exactly for the work done, `good` is being given a reward on top of it because we perfectly followed the process. If you start deciding personally when giving or not an extra reward because "lot of work was done", it's the beginning of the end of a well oiled process :)

**Comment by sereshqua** [comment]
@victornoel Victor, you know, in different projects people have different opinions :) I have talked to a developer in Portal project, and he + some other people think opposite and when I punished for things like a comment, they had objection about it and were not happy. Now I am curious what to do

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @sereshqua[/z](https://www.0crat.com/u/sereshqua)

**Comment by llorllale** [comment]
@victornoel I think there is room in the [policy](http://www.zerocracy.com/policy.html#30) for clarification.

**Comment by 0crat** [comment]
Order was finished, quality is "good": +35 point(s) just awarded to @scristalli[/z](https://www.0crat.com/u/scristalli)
